### PR TITLE
Refactoring trait name

### LINKED ISSUE #727: [style] Rename trait type names from `I$Name` to `$Name`
**Summary**

E.g: rename `ITable` to `Table`. There is no point to add `I` prefix before a trait.

In rust, A trait type has two usages: 

1. Template argument constrait: like `Where T: Table`, we can know Table is a trait as soon as we see it. 
2. Trait objects: like `Box<dyn Table>`, dyn trait requires exiplit dyn keyword, there is also no confusion at all.

For the reason above, trait named with `I$Name` is meaningless.

### LINKED ISSUE #744: [style] Rename trait type names from $Name to I$Name
**Summary**

Due to the discussion and PR from #740, the result is that I prefix is important for readability .

A codebase should be consistent. it’s important to keep the same style, please rename trait without I to I$Name

such as

* pub trait ResultExt
* pub trait ExprRewriter
* pub trait ExpressionVisitor
* pub trait ToErrorCodes
…

And write some lint rules in CI to disable trait without I prefix in codebase.

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Rename trait type names from I$Name to $Name

(after a force-push, I failed to re-open OR #740, thus I create a new PR)

## Changelog

Rename trait type names from I$Name to $Name

## Related Issues

Fixes #727 
Fixes #744 

## Test Plan

No extra ut/stateless_test

### CODE REVIEW

File: fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs [hunk 1]
```diff
@@ -14,7 +14,7 @@ use common_planners::CreateTablePlan;
 use common_planners::TableEngineType;
 use common_planners::TableOptions;
 use fuse_query::interpreters::InterpreterFactory;
-use fuse_query::optimizers::Optimizer;
+use fuse_query::optimizers::Optimizers;
```
Comment (jyizheng) [inline fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs:17]: What is the reason for this change?

File: fusequery/query/src/datasources/remote/store_client_provider.rs [hunk 2]
```diff
@@ -9,8 +9,8 @@ use common_exception::Result;
 use common_flights::StoreClient;
 
 #[async_trait::async_trait]
-pub trait IStoreClientProvider {
+pub trait TryGetStoreClient {
```
Comment (jyizheng) [inline fusequery/query/src/datasources/remote/store_client_provider.rs:12]: What if the trait is extended with more functions in the future?

**Review by ZhiHanZ — APPROVED** [review]
LGTM

File: fusequery/query/src/datasources/datasource.rs [hunk 3]
```diff
@@ -20,32 +20,14 @@ use crate::datasources::remote::RemoteDatabase;
 use crate::datasources::remote::RemoteFactory;
 use crate::datasources::remote::RemoteTable;
 use crate::datasources::system::SystemFactory;
-use crate::datasources::IDatabase;
-use crate::datasources::ITable;
-use crate::datasources::ITableFunction;
-
-#[async_trait::async_trait]
-pub trait IDataSource: Sync + Send {
```
Comment (jyizheng) [inline fusequery/query/src/datasources/datasource.rs:28]: Is this necessary to support different source files, e.g., csv, table, parquet?

**Comment by jyizheng** [comment]
LGTM. Though I left some comments, please ignore them if they are not relevant.

File: fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs (hunk 1 above)
Comment (dantengsky) [reply fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs:17]: There is a naming confliction between `trait IOptimizer` and `struct Optimizer` (before refactoring).

`struct Optimizer` holds a vector of IOptimizer (thus, Optimizer**s**), like this:
~~~
pub struct Optimizer {    optimizers: Vec<Box<dyn IOptimizer>> }
pub trait IOptimizer { ....}
~~~

To resolve the naming confliction, it seems to me that rename `struct Optimizer` to `struct Optimizers` and `trait IOptimizer`  to `trait Optimizer` is a reasonable approach. 

Any suggestions?

File: fusequery/query/src/datasources/remote/store_client_provider.rs (hunk 2 above)
Comment (dantengsky) [reply fusequery/query/src/datasources/remote/store_client_provider.rs:12]: At that time, we should add another trait or refactor this trait, it depends, IMHO.

File: fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs (hunk 1 above)
Comment (jyizheng) [reply fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs:17]: It makes sense to rename Optimizer to Optimizers since the field is named Optimizers.

File: fusequery/query/src/datasources/datasource.rs (hunk 3 above)
Comment (BohuTANG) [reply fusequery/query/src/datasources/datasource.rs:28]: Currently it is not going to support

File: fusequery/query/src/datasources/remote/store_client_provider.rs (hunk 2 above)
Comment (jyizheng) [reply fusequery/query/src/datasources/remote/store_client_provider.rs:12]: Fine with me.

File: fusequery/query/src/datasources/datasource.rs (hunk 3 above)
Comment (dantengsky) [reply fusequery/query/src/datasources/datasource.rs:28]: The name `DataSource` is somehow misleading, IMHO; The functionalities it provides are more like a catalog (e.g. CK's DatabaseCatalog) than a DataSource(especially in the context of Spark).

The reason for removing IDataSource and use a struct directly  :

> But datafuse have some conflicts after rename:
> trait IDataSource
> struct Datasource
> 
> If we rename IDataSource to DataSource, how to deal the struct name?
> 

_Originally posted by @BohuTANG in https://github.com/datafuselabs/datafuse/issues/727#issuecomment-854820280_


> Remove `IDataSource`, use `struct DataSource` directly? (replace `Arc<dyn IDatasource>` with `Arc<DataSource>`). 
> 
> Till we have multiple impls for an "IDataSource", and at that time we will have different names for impls naturally.

_Originally posted by @dantengsky in https://github.com/datafuselabs/datafuse/issues/727#issuecomment-854845547_

But, I am considering re-introduce the DataSouce/Catalog trait (but mainly for ease of testing), hopefully, there will be an PR in few days.

File: fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs (hunk 1 above)
Comment (dantengsky) [reply fusequery/benchmarks/nyctaxi/src/bin/nyctaxi.rs:17]: Oh, forget to mention that field `optimizers`  is renamed to `inner` as well.

**Comment by dantengsky** [comment]
> LGTM. Though I left some comments, please ignore them if they are not relevant.

Txs bro
